lib/cocoapods/installer/pod_source_installer.rb

```diff
@@ -90,7 +90,8 @@ def clean!
       #
       def lock_files!(file_accessors)
         return if local?
-        FileUtils.chmod('u-w', source_files(file_accessors))
+        unlocked_files = source_files(file_accessors).reject { |f| (File.stat(f).mode & 0o200).zero? }
```

Comment:
this seems like it would be slower to me, since it's 2 syscalls per file (1 to stat and one to lock), vs just 1 for only doing the `chmod`

Reply:
Here's the thing though - and I don't know why or under what circumstances - `fstat` seems much cheaper than `chmod` so in the case of an incremental install I measured a significant performance benefit.

Reply:
@igor-makarov should we cross reference if `unlock_files!` or `lock_files!` in `installer.rb` to use the list of generated pod targets instead?

```ruby
    # Unlocks the sources of the Pods.
    #
    def unlock_pod_sources
      pod_installers.each do |installer|
        pod_target = pod_targets.find { |target| target.pod_name == installer.name }
        installer.unlock_files!(pod_target.file_accessors)
      end
    end

    # Locks the sources of the Pods if the config instructs to do so.
    #
    def lock_pod_sources
      return unless installation_options.lock_pod_sources?
      pod_installers.each do |installer|
        pod_target = pod_targets.find { |target| target.pod_name == installer.name }
        installer.lock_files!(pod_target.file_accessors)
      end
    end
```

Change the `.find` to be against `generated_pod_targets` and if its not found then dont do anything?

Reply:
Something like:
```ruby
    def unlock_pod_sources
      pod_installers.each do |installer|
        pod_target = generated_pod_targets.find { |target| target.pod_name == installer.name }
        installer.unlock_files!(pod_target.file_accessors) if pod_target
      end
    end

    # Locks the sources of the Pods if the config instructs to do so.
    #
    def lock_pod_sources
      return unless installation_options.lock_pod_sources?
      pod_installers.each do |installer|
        pod_target = generated_pod_targets.find { |target| target.pod_name == installer.name }
        installer.lock_files!(pod_target.file_accessors) if pod_target
      end
    end
```

Reply:
This seems like a nice idea but I don't feel confident in that part of the code and it feels like a bigger change to me?

The `fstat` change is a local optimization so I feel better about it.

Reply:
Additionally, I feel like coupling it with the incremental install feature could introduce UB, for example, a user unlocking a file in Xcode and it never getting locked again because the project doesn't get regenerated.